### PR TITLE
Fix #10391 - incorrect cursor position after autocomplete

### LINKED ISSUE #10391: Completer places cursor in wrong position after dot complete (master)
## Description

Probably follow up after #10234 is needed. Any thoughts @dmonad?

## Reproduce

0. Using Python kernel
1. Enter `dict.` into a cell
2. Press <kbd>Tab</kbd>
3. Select any completion
4. See the cursor still remaining after the dot rather than at the end of the chosen method

completer-cursor-issue

## Expected behavior

The cursor caret lands after the chosen method i.e. `dict.clear|` (as it does in stable version) rather than `dict.|clear`

## Context

- Browser and version: Chrome
- JupyterLab version: 3.1.0a12 / master

### PR DESCRIPTION
**Pull request by dmonad**
## References

This PR fixes the issue described in #10391. 

## Code changes

The auto-complete plugin now explicitly sets the cursor position after the replacement.

Previously, we relied on the behavior of modeldb to adapt the cursor position correctly. The autocomplete plugin simply replaced text and the local position would be automatically updated to the position after the replacement. This is no longer possible in all cases because y-codemirror now updates the cursor positions after each change with an algorithm that is designed to work nicely with remote changes. We instead update the position to the position before the replacement, because that will lead to fewer conflicts when multiple peers insert content at the same time. 

## User-facing changes

None

## Backwards-incompatible changes

None

### CODE REVIEW

**Review by blink1073 — APPROVED** [review]
Thank you!

**Review by krassowski — APPROVED** [review]
Thank you!

**Comment by dmonad** [comment]
Thanks for the quick review! I fixed the tests, hope they run now.

**Comment by krassowski** [comment]
Could you describe I don't think the fix works now. The `dict.<tab>` example gives me `dict.|clear` again. What was the sequence the tests were failing on?

**Comment by krassowski** [comment]
Yup, not working after the changes that fixed the test:

not-working

Debug info:

```ts
patch = {start: 5, end: 5, value: "clear"};
cursorBeforeChange = 5
```

this means that the condition of `cursorBeforeChange < end && cursorBeforeChange >= start` will not be true and the fix will not be applied. In fact I, wonder if this condition will ever be true (that would require that cursor is simultaneously before `end` and after `start` which should never be true for well behaved patches?) - did you mean to make the condition the other way round? It would help if you add a comment on why this was needed.

Also, maybe it would be useful to add a test for `dict.|<tab>` -> `dict.clear|`?

**Comment by dmonad** [comment]
Thanks @krassowski for trying out the patch! I fixed the issue on by including the `end` as part of the range. I also added a test for autocompletion on an empty work (replace "" with "foobar" for example, which should cover the cases described in the original issue). 

> that would require that cursor is simultaneously before end and after start which should never be true for well behaved patches?

When performing autocomplete on the empty word (e.g. "" ⇒ "dict"), the cursor is within the replaced range. Although in this case `start == end`. Usually, the cursor would be after the replaced range (e.g. "di|" ⇒ "dict|"). 

This PR only fixes an edge case. When the cursor is within the replaced word (or directly after it), we expect that the cursor is after the replaced word after the change is performed. All other cases are handled by the existing cursor mapping approach.

**Comment by dmonad** [comment]
I rebased because the tests were failing. Now different tests that are unrelated to this PR are failing. I'm going to try rebasing again tomorrow.

File: packages/completer/test/handler.spec.ts [hunk 1]
```diff
@@ -320,5 +320,41 @@ describe('@jupyterlab/completer', () => {
...
+      const offset = handler.editor.getOffsetAt({ line, column });
+      model.cursor = { start: offset, end: offset };
+      // Make the completion, check its value and cursor position.
+      (completer.selected as any).emit(patch);
+      expect(editor.model.value.text).toBe(want);
+      expect(editor.getCursorPosition()).toEqual({
+        line,
+        column: column + 6
+      });
+      console.warn(editor.getCursorPosition());
```
Comment (krassowski) [inline packages/completer/test/handler.spec.ts:357]:
```suggestion
```

Comment (krassowski) [reply packages/completer/test/handler.spec.ts:357]: Or not? Like we have this in a couple of places but it is a bit distracting to me so I would personally prefer to avoid adding new logs in tests.

Comment (dmonad) [reply packages/completer/test/handler.spec.ts]: Agreed. Less is better. Since these logs are all over the place, I was thinking that some people use these for debugging. I simply copied the existing code so everything looks similar.

**Review by krassowski — APPROVED** [review]

**Comment by blink1073** [comment]
@meeseeksdev please backport to 3.1.x
